TFNetworkLayer.py

```diff
@@ -1107,53 +1111,218 @@ def batch_norm(self, data,
       tf.nn.batch_normalization()
       https://github.com/deepmind/sonnet/blob/master/sonnet/python/modules/batch_norm.py
     """
+    from operator import mul
+    from functools import reduce
+
     with tf.variable_scope("batch_norm"):
-      if masked_time:
-        x = data.get_placeholder_flattened(keep_dims=True)
-        mean, variance = tf.nn.moments(x, axes=[0], keep_dims=True)
-      else:
-        x = data.placeholder
-        mean, variance = tf.nn.moments(x, axes=data.get_axes(exclude_feature=True), keep_dims=True)
+      # Creating variables
+      fused = fused and not force_sample
+      # Perform batch_norm operations in NHWC format in order to
+      # keep consistency when saving/loading the variables.
+      # Not relevant when fused or masked_time.
+      input_data = data.copy() if fused or masked_time else data.copy_as_batch_spatial_major()
```

Comment:
This is not clear. Why do you make it batch-spatial major if not fused and not masked time?

Reply:
Also, this makes it inefficient, or not? And is not really needed?